apollo-testing-support/src/appleMain/kotlin/com/apollographql/apollo3/testing/runWithMainLoopApple.kt

```diff
@@ -26,23 +26,26 @@ import kotlin.coroutines.CoroutineContext
  * but for now that allows us to run integration tests against a mocked server
  */
 @OptIn(DelicateCoroutinesApi::class)
-actual fun <T> runWithMainLoop(context: CoroutineContext, block: suspend CoroutineScope.() -> T): T {
-  var result: Result<T>? = null
-  GlobalScope.launch (MainLoopDispatcher) {
-    result = kotlin.runCatching {
+actual fun runTest(
+    context: CoroutineContext,
+    before: suspend CoroutineScope.() -> Unit,
+    after: suspend CoroutineScope.() -> Unit,
+    block: suspend CoroutineScope.() -> Unit,
+) {
+  GlobalScope.launch(context + MainLoopDispatcher) {
+    before()
+    try {
       block()
+    } finally {
+      after()
```

Comment:
Can we keep a simple `runTest(block: suspend CoroutineScope.() -> T): T` ? It's important that exceptions happening inside the block are forwarded to the caller and not ignored. Getting back a return value might be useful too. And it's easy to implement before/after at the callsite:

```kotlin
  private fun runTestWithMockServer(block: (MockServer) -> Unit) {
    runTest {
      val mockServer = MockServer()
      try {
        block(mockServer)
      } finally {
        mockServer.stop()
      }
    }
  }
```

Reply:
I don't think we can.

If you want to call suspending functions from the test, then the test method in JS must return a `Promise`. There is no other way. Notice that in Js implementation, the return value is being re-casted to `dynamic`, to kinda make the compiler happy and allowing me to return that `Promise` instead of `Unit`.

Regarding the `before` and `after` parameters, there is another issue with asynchronicity. Because we've made `MockServer.url()` suspending, also the `@BeforeTest` method setting up the server and client must be suspending. So in JS world, it must return a `Promise`. But if it does, the ordering of things is different then one may expect. Because the testing framework does not wait for that promise to resolve before starting the actual test. See it [here](https://youtrack.jetbrains.com/issue/KT-19813#focus=Comments-27-3536253.0-0).

So that's the reasons why there are those 2 parameters, and why the `runTest` method looks how it looks.

Reply:
I understand `@BeforeTest` is not working in JS but `suspend fun` are still executed in order, right? If `runTest` is used to start the coroutine, then I think before/after can be run from this coroutine?

```kotlin
actual fun runTest(
    context: CoroutineContext,
    block: suspend CoroutineScope.() -> Unit,
): dynamic = GlobalScope.promise(context) {
  return block()
} 
```

And in the test:

```kotlin
  private fun runTestWithMockServer(block: suspend (MockServer) -> Unit) = runTest {
    val mockServer = MockServer()
    try {
      block(mockServer)
    } finally {
      mockServer.stop()
    }
  }
```

Is there a specific reason this wouldn't work? That means we can't get a return value from `runTest` which I guess is Ok. 

Side note: I wonder how the runner is waiting for the test to complete before launching other tests now :). Since there's no blocking call in JS, there has to be another way of scheduling all the tests